tests/NavigationTests.elm

```diff
@@ -0,0 +1,156 @@
+module NavigationTests exposing (..)
+
+import Expect
+import Html exposing (..)
+import Navigation
+import Test exposing (..)
+import TestContext exposing (..)
+
+
+type Msg
+    = UrlChange Navigation.Location
+    | PushUrl String
+    | ModifyUrl String
+    | Back Int
+    | Forward Int
+    | Load String
+    | Reload
+
+
+type alias Model =
+    { location : Navigation.Location, msgs : List Msg }
+
+
+init : Navigation.Location -> ( Model, Cmd Msg )
+init location =
+    ( { location = location, msgs = [] }, Cmd.none )
+
+
+initWithStringFlags : String -> Navigation.Location -> ( Model, Cmd Msg )
+initWithStringFlags flags location =
+    init location
+
+
+programUpdate : Msg -> Model -> ( Model, Cmd Msg )
+programUpdate msg model =
+    case msg of
+        UrlChange location ->
+            ( { model | location = location, msgs = msg :: model.msgs }, Navigation.modifyUrl location.pathname )
```

Comment:
Might be clearer to call the `msgs` field `urlChanges` instead since it doesn't record all messages.